ui/apps/platform/src/Components/CompoundSearchFilter/types.ts

```diff
@@ -0,0 +1,345 @@
+            filterChipLabel: 'Platform CVE ID';
+            searchTerm: 'CVE ID';
+            inputType: 'autocomplete';
+        };
+        DiscoveredTime?: {
+            displayName: 'Discovered Time';
+            filterChipLabel: 'Platform CVE Discovered Time';
+            searchTerm: 'CVE Created Time';
+            inputType: 'date-picker';
+        };
+        CVSS?: {
+            displayName: 'CVSS';
+            filterChipLabel: 'Platform CVE CVSS';
+            searchTerm: 'CVSS';
+            inputType: 'dropdown-slider';
+        };
+        Snoozed?: {
+            displayName: 'Snoozed';
+            filterChipLabel: 'Platform CVE Snoozed';
+            searchTerm: 'CVE Snoozed';
+            inputType: 'select';
+        };
+        Type?: {
+            displayName: 'Type';
+            filterChipLabel: 'Platform CVE Type';
+            searchTerm: 'CVE Type';
+            inputType: 'select';
+        };
+    };
+};
+
+export type ImageComponentSearchFilterConfig = {
+    displayName: 'Image Component';
+    searchCategory: 'IMAGE_COMPONENTS';
+    attributes: {
+        Name?: {
+            displayName: 'Name';
+            filterChipLabel: 'Image Component Name';
+            searchTerm: 'Component';
+            inputType: 'autocomplete';
+        };
+        Source?: {
+            displayName: 'Source';
+            filterChipLabel: 'Image Component Source';
+            searchTerm: 'Component Source';
+            inputType: 'select';
+        };
+        Version?: {
+            displayName: 'Version';
+            filterChipLabel: 'Image Component Version';
+            searchTerm: 'Component Version';
+            inputType: 'text';
+        };
+    };
+};
+
+export type NodeComponentSearchFilterConfig = {
+    displayName: 'Node Component';
+    searchCategory: 'NODE_COMPONENTS';
+    attributes: {
+        Name?: {
+            displayName: 'Name';
+            filterChipLabel: 'Node Component Name';
+            searchTerm: 'Component';
+            inputType: 'autocomplete';
+        };
+        Source?: {
+            displayName: 'Source';
+            filterChipLabel: 'Node Component Source';
+            searchTerm: 'Component Source';
+            inputType: 'select';
+        };
+        Version?: {
+            displayName: 'Version';
+            filterChipLabel: 'Node Component Version';
+            searchTerm: 'Component Version';
+            inputType: 'text';
+        };
+    };
+};
+
+export type CompoundSearchFilterConfig = {
+    Image?: ImageSearchFilterConfig;
+    Deployment?: DeploymentSearchFilterConfig;
+    Namespace?: NamespaceSearchFilterConfig;
+    Cluster?: ClusterSearchFilterConfig;
+    Node?: NodeSearchFilterConfig;
+    ImageCVE?: ImageCVESearchFilterConfig;
+    NodeCVE?: NodeCVESearchFilterConfig;
+    PlatformCVE?: PlatformCVESearchFilterConfig;
+    ImageComponent?: ImageComponentSearchFilterConfig;
+    NodeComponent?: NodeComponentSearchFilterConfig;
+};
```

Comment:
Yes, thanks for the suggestions—they're worth exploring. I've incorporated some of your ideas in the PR. We can establish a high-level model similar to your example for each JavaScript configuration object. From there, we could develop specific types for each entity configuration, incorporating optional attributes for practical implementation.

Reply:
I've updated the util function to be more reusable as well. Please take a look at that and see if the Typescript makes sense. I'm not a Typescript guru 😬 but after experimenting, this seems to work well for what we're trying to do

Reply:
Yeah this looks great, and works exactly as you would expect from the calling side -- we get a subset of attributes and TypeScript correctly knows what keys exist on the results.

I usually try to avoid `as Type` whenever possible, but it looks like in this case it isn't possible! Even the [TypeScript gurus](https://stackoverflow.com/a/73083427) [note](https://stackoverflow.com/a/74553065) that this is a [limitation](https://stackoverflow.com/a/58986422) in the type system. I like the quote in the last link:
> I think the type assertion really isn't all that bad; it's meant specifically for situations in which you are smarter than the compiler... and this seems to be one of those times.

IMO using plain old `forEach` is more readable than `reduce`:
```ts
    const attributes = {} as Pick<T['attributes'], K>;

    selectedAttributes.forEach((key) => {
        attributes[key] = config.attributes[key as string];
    });

    return {
        ...config,
        attributes,
    };
```
but for a function this short, having unit tests (🙌 ) cover the safety net we don't get from the type system.

---

It is interesting to see all of the complex things you can do with TS, but given the seemingly simple problem of "I have an object with known keys, give me a new object with a selection of those keys removed" and it can't do it without a bit of cheating.